src/aiai_eval/task.py

```diff
@@ -54,7 +82,567 @@ def evaluate(self, model_id: str) -> Dict[str, dict]:
+        test: Dataset,
+        rng: np.random.Generator,
+        model_config: ModelConfig,
+        num_iter: int,
+    ) -> Dict[str, Dict[str, float]]:
+        """Evaluate a PyTorch or JAX model.
+
+        Args:
+            model_dict (dict):
+                The model dictionary, with keys "model" and "tokenizer".
+            test (Dataset):
+                The test dataset.
+            rng (np.random.Generator):
+                The random number generator, used to generate bootstrapped versions of
+                the test dataset.
+            model_config (ModelConfig):
+                The model configuration.
+            num_iter (int):
+                The number of bootstrapped samples of the test dataset to use.
+
+        Returns:
+            dict:
+                The keys in the dict are 'raw' and 'total', with all the raw scores in
+                the first dictionary and the aggregated scores in the second.
+        """
+        # Extract the model and tokenizer
+        model = model_dict["model"]
+        tokenizer = model_dict["tokenizer"]
+
+        # Log the number of parameters in the model
+        num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
+        logger.info(f"Number of model parameters: {num_params:,}")
+
+        # Preprocess the datasets
+        try:
+            params = dict(
+                framework="pytorch",
+                config=model.config,
+                tokenizer=tokenizer,
+                dataset_task=self.dataset_task,
+            )
+            test = self._preprocess_data(test, **params)
+        except ValueError:
+            raise InvalidEvaluation("Preprocessing of the dataset could not be done.")
+
+        # If we are testing then truncate the test set
+        if self.evaluation_config.testing:
+            test = Dataset.from_dict(test[:128])
+
+        # Get bootstrapped datasets
+        tests = [
+            Dataset.from_dict(test[rng.integers(0, len(test), len(test))])
+            for _ in range(num_iter)
+        ]
+
+        # Set up progress bar
+        if self.evaluation_config.progress_bar:
+            itr = tqdm(range(num_iter), desc="Benchmarking")
+        else:
+            itr = range(num_iter)
+
+        # Load the data collator
+        data_collator = self._load_data_collator(tokenizer)
+
+        scores = []
+        for idx in itr:
+            while True:
+                test_itr_scores = self._evaluate_pytorch_jax_single_iteration(
+                    idx=idx,
+                    model_config=model_config,
+                    tests=tests,
+                    data_collator=data_collator,
+                )
+                # If the iteration was successful then break the while-loop
+                if isinstance(test_itr_scores, dict):
+                    break
+
+                # Otherwise we encountered an error
+                else:
+                    raise InvalidEvaluation(str(test_itr_scores))
```

Comment:
Maybe this needs to be a bit more descriptive?

Reply:
This is like a last ditch effort to throw an exception, in case we have not successfully raised a more meaningful one earlier in `_evaluate_pytorch_jax_single_iteration`. Do you have a suggestion? I am not entirely sure how to do it more specific, besides some fluff around the error message, i.e. "An unknown error occurred doing evaluation of a single iteration .... {err_msg}..."

Reply:
True. I think the fluff _would_ make it a bit more descriptive though, especially if it ends with something like "(...), the test scores were the following: {test_itr_scores}".

Reply:
Coolio, ill add it.